File: karate-chop/java/src/main/java/com/codekata/karatechop/IterativeBinarySearch.java
package com.codekata.karatechop;

public final class IterativeBinarySearch implements BinarySearch {

    @Override
    public int chop(int numberToFind, int[] inputNumbers) {

        return binarySearch(inputNumbers, numberToFind, 0, inputNumbers.length - 1);
    }

    private int binarySearch(int[] inputNumbers, int numberToFind, int imin, int imax) {

        while (imin <= imax) {
            int imid = midpoint(imin, imax);

            if(inputNumbers[imid] == numberToFind) {
                return imid;
            } else if (inputNumbers[imid] < numberToFind) {
                imin = imid + 1;
            } else {
                imax = imid - 1;
            }
        }

        return NOT_FOUND;
    }

    private int midpoint(int imin, int imax) {
        return (imin + imax) / 2;
    }
}

File: karate-chop/js/components/IterativeBinarySearch.js
"use strict";

class IterativeBinarySearch {

    chop(numberToFind, inputNumbers) {

        function binarySearch(inputNumbers, numberToFind, imin, imax) {

            function midpoint(imin, imax) {
                return Number.parseInt((imin + imax) / 2);
            }

            while (imin <= imax) {
                let imid = midpoint(imin, imax);

                if (inputNumbers[imid] === numberToFind) {
                    return imid;
                } else if (inputNumbers[imid] < numberToFind) {
                    imin = imid + 1;
                } else {
                    imax = imid - 1;
                }
            }

            return -1;
        }

        return binarySearch(inputNumbers, numberToFind, 0, inputNumbers.length - 1);
    }
}

module.exports = IterativeBinarySearch;
